ui/src/preview/preview-view.tsx

```diff
@@ -118,48 +131,93 @@ export const PreviewView: React.FC<PreviewProps> = (props: PreviewProps) => {
-                    );
+    const [designPublicationMode, setDesignPublicationMode] =
+        React.useState<DesignPublicationMode>('PUBLISH_CHANGES');
 
-                    return candidateIsStaged
-                        ? {
-                              ...candidate,
-                              stage: PublicationStage.STAGED,
-                          }
-                        : candidate;
-                });
+    const onChangeDesignPublicationMode = (newMode: DesignPublicationMode) => {
+        setDesignPublicationMode(newMode);
+        if (newMode === 'PUBLISH_CHANGES') {
+            // TODO GVT-2421: No ability to validate when moving changes to main yet; just pretend everything is OK
+            setShowValidationStatusSpinner(false);
+        }
+    };
+
+    const canRevertChanges =
+        props.layoutContext.designId === undefined || designPublicationMode === 'PUBLISH_CHANGES';
+
+    const [mapDisplayTransitionSide, setMapDisplayTransitionSide] =
+        React.useState<MapDisplayTransitionSide>('WITH_CHANGES');
+
+    const hasDesignDraftsInDesign =
+        useLoader(
+            () =>
+                props.layoutContext.designId === undefined ||
+                designPublicationMode === 'PUBLISH_CHANGES'
+                    ? Promise.resolve(false)
+                    : getPublicationCandidates(props.layoutContext.designId, 'DRAFT').then(
+                          (candidates) => candidates.length > 0,
+                      ),
+            [props.changeTimes, props.layoutContext.designId, designPublicationMode],
+        ) ?? false;
+
+    useTwoPartEffectWithStatus(
+        () =>
+            getPublicationCandidates(
+                props.layoutContext.designId,
+                designPublicationMode === 'PUBLISH_CHANGES' ? 'DRAFT' : 'OFFICIAL',
+            ),
+        (candidates) => {
+            const candidatesWithUpdatedStage = candidates.map((candidate) => {
+                const candidateIsStaged = props.stagedPublicationCandidateReferences.some(
+                    (stagedCandidate) => candidateIdAndTypeMatches(stagedCandidate, candidate),
+                );
+
+                return candidateIsStaged
+                    ? {
+                          ...candidate,
+                          stage: PublicationStage.STAGED,
+                      }
+                    : candidate;
+            });
 
-                setPublicationCandidates(candidatesWithUpdatedStage);
-                props.setStagedPublicationCandidateReferences(candidatesWithUpdatedStage);
+            setPublicationCandidates(candidatesWithUpdatedStage);
+            props.setStagedPublicationCandidateReferences(candidatesWithUpdatedStage);
 
-                setShowPreview(true);
-            }),
-        [props.changeTimes],
+            setShowPreview(true);
+        },
+        [props.changeTimes, props.layoutContext.designId, designPublicationMode],
```

Comment:
Julkaisuehdokkaiden haku ei ole isommin muuttunut itsessään, paitsi että vääränmuotoinen useLoader()in käyttö on vaihdettu pois. Haku muotoa `useLoader(() => getFoo().then((result) => setFoo(result))` jättää tilaan väärän foon, jos getFoo()t palaavat palvelimelta väärässä järjestyksessä.

Tämä siis korrektiussyistä. Koodin saattaisi silti saada selkeämmäksi refaktoroimalla tämän takaisin useLoader()in kutsuksi mutta oikeaoppisesti (ts. ottamalla vaan sen palautusarvon, ja hoitamalla lataukseen liittyvät tilanhallinnat erikseen).

Reply:
Hyvä huomio ja korjaus tosiaan. Tämä sai miettimään, että tuollaisten tilanteiden aiheuttamat ongelmat eivät välttämättä rajoitu pelkästään näihin kaksitasoisiin hakuihin (yksittäisetkin haut voivat tulla bäkkäriltä takaisin väärässä järjestyksessä, ja mielestäni olen ainakin itse joskus sellaisia korjannutkin.) Meillä saattaisi siis olla tilausta jonkinlaiselle `useLoader`-variantille, jonka saisi canceloimaan aiemmat kutsut jotka eivät ole vielä palanneet bäkkäriltä.